doc/heps/hep2.md

```diff
@@ -0,0 +1,370 @@
+        ...
+      ```
+
+    - The `warn` function from the `warnings` only emits a run-time warning. It is however more versatile than `@deprecated` and can be used, for example, to deprecate a keyword parameter.
+
+    ```python
+    from warnings import warn
+
+    def foo(a, b=None):
+      warn(
+        "The keyword parameter `b` of function foo is deprecated since version 1.1.1 and will be removed in a future release.",
+        category=FutureWarning, stacklevel=2,
+      )
+      ...
+    ```
+
+  - The warning message:
+
+    - **Must** indicate that the feature is deprecated and is going to be removed in a future version.
+    - **Must** suggest replacement APIs, if applicable.
+    - Can indicate in which version the feature was deprecated.
+    - Can indicate before which version the feature is going to be remove, however, maintainers should ensure that the deprecation period (see below) is not going to be too to short, and that no obsolete warning will be released (e.g. feature announced to be removed in version 1.1 but is still present in 1.1).
+
+  - The warning type:
+  
+    - **Must** be a subclass of `DeprecationWarning`, `FutureWarning`, or `PendingDeprecationWarning`.
+    - **Must** be defined based on this approach (see the Appendix for more details):
+
+      1. Start by emitting a `DeprecationWarning`, to inform Library Developers, and some Data Analysts.
+      2. After a few months, upgrade to a `FutureWarning`, to inform all users.
+
+      Exceptions are allowed but should in practice be very rare:
+
+      - Start by emitting a `PendingDeprecationWarning` for deprecations that might be reverted or are expected to be extremely noisy.
+      - Don't upgrade to a `FutureWarning` when the feature removed is meant for Library Developers only.
+      - Emit directly a `FutureWarning` when the feature removed is meant for Data Analysts only.
+
+  - `stacklevel`, or `skip_file_prefixes` starting from Python 3.12, **must** be set when calling `warnings.warn` so the warning message references the right deprecated code line. Projects can implement a utility like the private Pandas' function [`find_stack_level`](https://github.com/pandas-dev/pandas/blob/b8a4691647a8850d681409c5dd35a12726cd94a1/pandas/util/_exceptions.py#L34) to automatically infer the right `stacklevel` number. Python 3.12 added the `skip_file_prefixes` to `warnings.warn` that could be used in place of setting `stacklevel` manually or with `find_stack_level`, with e.g. `skip_file_prefixes=os.path.dirname(importlib.util.find_spec(<package_name>).origin)`.
+
+- [ ] Documentation: all usage of the deprecated feature **must** be removed from the documentation, except from the section that serves as the reference API. Special cases (e.g. major API change) will need extra documentation. When the deprecation involves the user having to change their code in a significant way (typically in a major release), maintainers should consider writing a migration guide.
```

Comment:
To me, it's a bit unclear what is meant by `except from the section that serves as the reference API`. Shouldn't we also put a message in the reference API that it is going to be deprecated? This would be helpful if/when we move to versioned docs.

Reply:
Great catch thanks! This was in the original commit and got lost over edits. I'll re-add this bit:

> Docstring (if applicable): the docstring of the deprecated feature must be updated to mention the deprecation, when possible the [deprecated](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated) Sphinx directive should be used.

Reply:
Done in https://github.com/holoviz/holoviz/pull/388/commits/1712fa4353e1be7e1fbf71bf7bcf90b02847bec2